### ISSUE TITLE
explicit_auto_deref gives bad advice for lazy_static references

### ISSUE THREAD

**Issue by eric-seppanen** [rust-lang/rust-clippy#9777]
### Summary

In 1.65, clippy warns about code that it was fine with in 1.64. The suggested fix does not compile.

When encountering an explicit dereference of a "ref" struct generated by `lazy_static!`, it provides a suggested fix that doesn't work:
```
warning: deref which would be done by auto-deref
  --> src/lib.rs:12:9
   |
12 |     foo(*CONFIG);
   |         ^^^^^^^ help: try this: `CONFIG`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
   = note: `#[warn(clippy::explicit_auto_deref)]` on by default
```
If I apply that fix then the code no longer compiles:
```
error[E0308]: mismatched types
  --> src/lib.rs:12:9
   |
12 |     foo(CONFIG);
   |     --- ^^^^^^
   |     |   |
   |     |   expected `&str`, found struct `CONFIG`
   |     |   help: consider borrowing here: `&CONFIG`
   |     arguments to this function are incorrect
   |
note: function defined here
  --> src/lib.rs:9:4
   |
9  | fn foo(_: &'static str) {}
   |    ^^^ ---------------
```

The "help" is correct: `&CONFIG` works (with no lints).

### Reproducer

I tried this code:

```rust
fn read_config() -> &'static str {
    "abc"
}

lazy_static::lazy_static! {
    pub static ref CONFIG: &'static str = read_config();
}

fn foo(_: &'static str) {}

pub fn bar() {
    foo(*CONFIG);
}
```

I expected to see this happen:

Either there should be no lint, or the suggested fix should still compile.

Instead, this happened:

The suggest fix does not compile.

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0

It also fails on 1.67.0-nightly (11ebe6512 2022-11-01).
```
